File: go/problem/zigzag_conversion.go
package problem

// input: PAYPALISHIRING
// output: PAHNAPLSIIGYIR
func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}
	tmp := make([]string, numRows)

	rowIndex := 0
	down := false

	for _, char := range s {
		tmp[rowIndex] += string(char)

		if rowIndex == 0 {
			down = true
		} else if rowIndex == numRows-1 {
			down = false
		}

		if down {
			rowIndex++
		} else {
			rowIndex--
		}
	}

	result := ""
	for _, str := range tmp {
		result += str
	}
	return result
}

File: rust/tests/zigzag_conversion.rs
use rust::problems::zigzag_conversion::Solution;

#[cfg(test)]
mod tests {
    use super::*;

    struct TestCase {
        input: &'static str,
        num_rows: i32,
        expected: &'static str,
    }

    fn test_cases() -> Vec<TestCase> {
        vec![
            TestCase {
                input: "PAYPALISHIRING",
                num_rows: 3,
                expected: "PAHNAPLSIIGYIR",
            },
            TestCase {
                input: "PAYPALISHIRING",
                num_rows: 4,
                expected: "PINALSIGYAHRPI",
            },
        ]
    }

    #[test]
    fn test_convert() {
        for (i, case) in test_cases().iter().enumerate() {
            let result = Solution::convert(case.input.to_string(), case.num_rows);
            assert_eq!(
                result, case.expected,
                "convert failed at case {}: input='{}', num_rows={}",
                i, case.input, case.num_rows
            );
            let result = Solution::convert_v2(case.input.to_string(), case.num_rows);
            assert_eq!(
                result, case.expected,
                "convert_v2 failed at case {}: input='{}', num_rows={}",
                i, case.input, case.num_rows
            );
        }
    }
}
